retry/retry.go

```diff
@@ -0,0 +1,142 @@
+// It does not itself invoke the action to be retried, but
+// is intended to be used in a retry loop.
+//
+// The basic usage is as follows:
+//
+//	for a := someStrategy.Start(); a.Next(); {
+//		try()
+//	}
+//
+// See examples for details of suggested usage.
+package retry
+
+import (
+	"time"
+
+	"github.com/juju/utils/clock"
+)
+
+// Strategy is implemented by types that represent a retry strategy.
+type Strategy interface {
+	NewTimer(now time.Time) Timer
+}
+
+// Timer represents a source of timing events for a retry strategy.
+type Timer interface {
+	// NextSleep returns the length of time to sleep
+	// before the next retry. If no more attempts should
+	// be made it should return false, and the returned
+	// duration will be ignored.
+	//
+	// Note that NextSleep is called once after
+	// each iteration has completed, assuming the
+	// retry loop is continuing.
+	NextSleep(now time.Time) (time.Duration, bool)
+}
+
+// Attempt represents a running retry attempt.
+type Attempt struct {
+	clock clock.Clock
+	stop  <-chan struct{}
+	timer Timer
+
+	// nextSleep holds the next sleep duration. This
+	// is valid only when known is true.
+	nextSleep time.Duration
+
+	// count holds the iteration count.
+	count int
+
+	// known holds whether nextSleep and running are known.
+	known bool
+
+	// running holds whether the attempt is still going.
+	running bool
+
+	// stopped holds whether the attempt has been stopped.
+	stopped bool
+}
+
+// Start begins a new sequence of attempts for the given strategy. If
+// clk is nil, clock.WallClock will be used. If a value is received on
+// stop while waiting, the attempt will be aborted.
+func Start(strategy Strategy, clk clock.Clock, stop <-chan struct{}) *Attempt {
+	if clk == nil {
+		clk = clock.WallClock
+	}
+	return &Attempt{
+		clock:   clk,
+		stop:    stop,
+		timer:   strategy.NewTimer(clk.Now()),
+		running: true,
+		known:   true,
+	}
+}
+
+// Next reports whether another attempt should be made, waiting as
+// necessary until it's time for the attempt. It always returns true the
+// first time it is called unless a value is received on the stop
+// channel - we are guaranteed to make at least one attempt unless
+// stopped.
+func (a *Attempt) Next() bool {
+	if !a.More() {
+		return false
+	}
+	if a.nextSleep == 0 {
+		// We're not going to sleep for any length of time,
+		// so guarantee that we respect the stop channel. This
+		// ensures that we make no attempts if Next is called
+		// with a value available on the stop channel.
+		select {
+		case <-a.stop:
+			a.stopped = true
+			a.running = false
+			return false
+		default:
+			a.known = false
+			a.count++
+			return true
+		}
+	}
+	select {
+	case <-a.clock.After(a.nextSleep):
+		a.known = false
+		a.count++
+	case <-a.stop:
+		a.running = false
+		a.stopped = true
+	}
+	return a.running
+}
+
+// More reports whether the next call to Next will sleep again to wait
+// for another attempt. It does not itself block.
```

Comment:
This comment is really confusing. The firs thing it says is that it'll report if Next will sleep... .but that's an implementation detail.  What it reports is whether or not the attempt will retry.    I think that also indicates the name is not very good.  I don't know what More() means in the course of a an Attempt.  If it was called WillRetry, it would be very clear.